x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlUpgradeModeActionFilter.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.xpack.core.ml.action.UpdateModelSnapshotAction;
+import org.elasticsearch.xpack.core.ml.action.UpdateProcessAction;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * {@link MlUpgradeModeActionFilter} disallows certain actions if the cluster is currently in upgrade mode.
+ *
+ * Disallowed actions are the ones which can access/alter the state of ML internal indices.
+ */
+class MlUpgradeModeActionFilter extends ActionFilter.Simple {
+
+    private static final Set<String> ACTIONS_DISALLOWED_IN_UPGRADE_MODE =
+        Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+            PutJobAction.NAME,
+            UpdateJobAction.NAME,
+            DeleteJobAction.NAME,
+            OpenJobAction.NAME,
+            FlushJobAction.NAME,
+            CloseJobAction.NAME,
+            PersistJobAction.NAME,
+
+            FinalizeJobExecutionAction.NAME,
+            PostDataAction.NAME,
+
+            RevertModelSnapshotAction.NAME,
+            UpdateModelSnapshotAction.NAME,
+            DeleteModelSnapshotAction.NAME,
+
+            PutDatafeedAction.NAME,
+            UpdateDatafeedAction.NAME,
+            DeleteDatafeedAction.NAME,
+            StartDatafeedAction.NAME,
+            StopDatafeedAction.NAME,
+
+            PutFilterAction.NAME,
+            UpdateFilterAction.NAME,
+            DeleteFilterAction.NAME,
+
+            PutCalendarAction.NAME,
+            UpdateCalendarJobAction.NAME,
+            PostCalendarEventsAction.NAME,
+            DeleteCalendarAction.NAME,
+            DeleteCalendarEventAction.NAME,
+
+            UpdateProcessAction.NAME,
+            KillProcessAction.NAME,
+
+            DeleteExpiredDataAction.NAME,
+
+            ForecastJobAction.NAME,
+            DeleteForecastAction.NAME,
+
+            PutDataFrameAnalyticsAction.NAME,
+            DeleteDataFrameAnalyticsAction.NAME,
+            StartDataFrameAnalyticsAction.NAME,
+            StopDataFrameAnalyticsAction.NAME,
+
+            PutTrainedModelAction.NAME,
+            DeleteTrainedModelAction.NAME
+        )));
+
+    private final AtomicBoolean isUpgradeMode = new AtomicBoolean();
+
+    MlUpgradeModeActionFilter(ClusterService clusterService) {
+        Objects.requireNonNull(clusterService);
+        clusterService.addListener(this::setIsUpgradeMode);
+    }
+
+    @Override
+    protected boolean apply(String action, ActionRequest request, ActionListener<?> listener) {
+        if (isUpgradeMode.get() && ACTIONS_DISALLOWED_IN_UPGRADE_MODE.contains(action)) {
+            throw new ElasticsearchStatusException(
+                "Cannot perform {} action while upgrade mode is enabled", RestStatus.TOO_MANY_REQUESTS, action);
+        }
+        return true;
+    }
+
+    @Override
+    public int order() {
+        return 666;
```

Comment:
Please add a comment here to say that the key requirement is that this filter operates after the security action filter so that an unauthorised user gets told they're unauthorised instead of finding out whether upgrade mode is enabled.

It's not hard to achieve this as security's order is `Integer.MIN_VALUE` so any number greater than that would work.  So the actual number is not that important, but the reasoning behind it is.

Reply:
Done (changed to `MAX_VALUE` to avoid magic numbers).

Actually the only values I've seen returned by this method are `MIN_VALUE` and `0` so I was not sure what is the right value for this filter. I guess it can be `MAX_VALUE` so that it is performed as last.